### PR TITLE
dsl: Add Real, Imag, and Conj operators

### PR DESCRIPTION
**Pull request by EdCaunt**
Also fixes a bug where equations of the form:
```python
eq = Eq(g, Re(f+1))
```
where `f` is complex would result in an
```C
float r0 = f[x + 1] + 1;
```
rather than
```C
float _Complex r0 = f[x + 1] + 1;
```
This was because the dtype of the RHS of an equation was not being checked correctly in some cases.

### CODE REVIEW

File: devito/finite_differences/differentiable.py [hunk 1]
```diff
@@ -645,6 +645,39 @@ def __str__(self):
...
+class ComplexPart(Differentiable, sympy.core.function.Application):
+    """Abstract class for `Re` or `Im` of an expression"""
+
+    def __new__(cls, *args, **kwargs):
+        if len(args) != 1:
+            raise ValueError(f"{cls.__name__} is constructed with exactly one arg;"
+                             f" {len(args)} were supplied.")
+
+        # Diffify any Add, Mul, etc which might be in the expression
+        new_args = (diffify(args[0]),)
```
Comment (mloubout) [inline devito/finite_differences/differentiable.py:657]: This should not be needed,  `diffify` is not meant for corner cases induced by sympy's internal hardcoding of some classes like Add and Mul in some places. `diffify` should technically only be used with the compiler pass not for user level objects.

File: tests/test_symbolics.py [hunk 2]
```diff
@@ -818,3 +818,137 @@ def test_assumptions(self, op, expr, assumptions, expected):
...
+        with switchconfig(language=language):
+            Operator(eqs)()
+
+    def test_devito_print(self):
+        f, _, _ = self.setup_basic(np.complex64)
+
+        assert str(Real(f)) == 'Real(f(x))'
+        assert str(Imag(f)) == 'Imag(f(x))'
+
+    @pytest.mark.parametrize('language', ['C', 'CXX', 'CXXopenmp'])
```
Comment (mloubout) [inline tests/test_symbolics.py:843]: If you just change the CI you won't need this parametrization and only check `configuration['language']` for the assert (and no need for switchconfig)

**Review by mloubout — COMMENTED** [review]
Some testing/nitpicky comments but looking good.

File: devito/passes/iet/languages/C.py [hunk 3]
```diff
@@ -55,3 +55,15 @@ class CPrinter(BasePrinter, C99CodePrinter):
 
     def _print_ImaginaryUnit(self, expr):
         return '_Complex_I'
+
+    def _print_Real(self, expr):
+        return (f'{self.func_prefix(expr)}real{self.func_literal(expr).lower()}'
```
[the remark this answers is not in the document]
Comment (EdCaunt) [reply devito/passes/iet/languages/C.py:60]: Huh, not sure where that came from. Good spot

File: devito/passes/iet/languages/C.py [hunk 4]
```diff
@@ -55,3 +55,15 @@ class CPrinter(BasePrinter, C99CodePrinter):
 
     def _print_ImaginaryUnit(self, expr):
         return '_Complex_I'
+
+    def _print_Real(self, expr):
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/passes/iet/languages/C.py:59]: You can still `_print_ComplexPart` and add a special case for Conj, that would make it easier to extend if necessary.

File: devito/finite_differences/differentiable.py (hunk 1 above)
Comment (EdCaunt) [reply devito/finite_differences/differentiable.py:658]: It's needed here to guarantee that the arg has a dtype attribute, otherwise the CSE can't find the correct dtype and reverts back to float32, which results in values getting cast where they shouldn't be

Comment (mloubout) [reply devito/finite_differences/differentiable.py:658]: You have an MFE for it? If `diffify` was needed for  a new `Differentiable` object it would be a huge problem throughout.

Comment (EdCaunt) [reply devito/finite_differences/differentiable.py:658]: It is a new behaviour specific to `ComplexPart` as it changes the type from complex to real, which was previously not possible in Devito

**Review by mloubout — APPROVED** [review]

File: devito/ir/iet/visitors.py [hunk 5]
```diff
@@ -423,7 +423,10 @@ def visit_PointerCast(self, o):
             if o.flat is None:
                 shape = ''.join(f"[{self.ccode(i)}]" for i in o.castshape)
                 rshape = f'(*){shape}'
-                lvalue = c.Value(cstr, f'(*{self._restrict_keyword} {v}){shape}')
+                if shape:
+                    lvalue = c.Value(cstr, f'(*{self._restrict_keyword} {v}){shape}')
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/ir/iet/visitors.py:427]: Just nicer code avoiding un-necessary parenthesis with no shape (i e size 1 array like in norm).

File: devito/passes/clusters/cse.py [hunk 6]
```diff
@@ -78,7 +78,8 @@ def cse(cluster, sregistry=None, options=None, **kwargs):
     if cluster.is_fence:
         return cluster
 
-    make = lambda: CTemp(name=sregistry.make_name(), dtype=dtype)
+    make_dtype = lambda e: np.promote_types(e.dtype, dtype).type
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/passes/clusters/cse.py:81]: You could but likely would need a try except like above since the cluster.dtype can be a non-numpy. My be useful to wrap the whole thing in a utility for safety, `e` might not have a numpy dtype too now that I think of it.

File: devito/ir/iet/visitors.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/ir/iet/visitors.py:427]: nitpicking could use a small comment because at first it tricked my eyes

File: devito/finite_differences/differentiable.py [hunk 7]
```diff
@@ -645,6 +645,47 @@ def __str__(self):
     __repr__ = __str__
 
 
+class ComplexPart(Differentiable, sympy.core.function.Application):
+    """Abstract class for `Real`, `Imag`, or `Conj` of an expression"""
+    _name = None
+
+    def __new__(cls, *args, **kwargs):
+        if len(args) != 1:
+            raise ValueError(f"{cls.__name__} is constructed with exactly one arg;"
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py:654]: I'd rewrite it as 

* "is constructed with" -> "expects";
* "were supplied" -> "were supplied instead"

File: devito/passes/clusters/aliases.py [hunk 8]
```diff
@@ -1354,7 +1354,7 @@ def cost(self):
         # Not just the sum for the individual items' cost! There might be
         # redundancies, which we factor out here...
         counter = generator()
-        make = lambda: Symbol(name='dummy%d' % counter(), dtype=np.float32)
+        make = lambda _: Symbol(name='dummy%d' % counter(), dtype=np.float32)
```
Comment (FabioLuporini) [inline devito/passes/clusters/aliases.py:1357]: why?

Comment (mloubout) [reply devito/passes/clusters/aliases.py:1357]: '_cse'no expect `make` to take an arg.

**Review by FabioLuporini — APPROVED** [review]
GTG
